### PR TITLE
Per-rpc host name rewriting for the authority header

### PR DESCRIPTION
**Pull request by kannanjgithub**
Per-rpc host name rewriting for the authority header based on the selected route.

### CODE REVIEW

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java [hunk 1]
```diff
@@ -217,6 +221,9 @@ public void start(Listener2 listener) {
     ldsResourceName = XdsClient.canonifyResourceName(ldsResourceName);
     callCounterProvider = SharedCallCounterMap.getInstance();
     resolveState = new ResolveState(ldsResourceName);
+    ServerInfo serverInfo = xdsClient.getServerInfo(ldsResourceName);
+    isTrustedXdsServer = serverInfo != null && serverInfo.isTrustedXdsServer();
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsNameResolver.java:225]: "Is trusted" should not leak out of the xds client. We want to "centrally" manage the trust. Having security checks strewn randomly through the code greatly increases the risk of a security bug.

Per the gRFC, this should only be used when validating/converting the route resources, which means `XdsRouteConfigureResource` and `XdsListenerResource`. If it isn't set in `XdsResourceType.Args.getServerInfo()`, we don't allow autoHostRewrite to be set.

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java [hunk 2]
```diff
@@ -465,14 +472,20 @@ public Result selectConfig(PickSubchannelArgs args) {
...
       class ClusterSelectionInterceptor implements ClientInterceptor {
         @Override
         public <ReqT, RespT> ClientCall<ReqT, RespT> interceptCall(
             final MethodDescriptor<ReqT, RespT> method, CallOptions callOptions,
             final Channel next) {
-          final CallOptions callOptionsForCluster =
+          CallOptions callOptionsForCluster =
               callOptions.withOption(CLUSTER_SELECTION_KEY, finalCluster)
                   .withOption(RPC_HASH_KEY, hash);
+          if (isTrustedXdsServer
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsNameResolver.java:484]: This is not safe with federation. The listener and routes could come from different control planes. In fact, different routes could come from different control planes. You need to check for this specific resource.

File: xds/src/main/java/io/grpc/xds/XdsNameResolver.java [hunk 3]
```diff
@@ -217,6 +221,9 @@ public void start(Listener2 listener) {
     ldsResourceName = XdsClient.canonifyResourceName(ldsResourceName);
     callCounterProvider = SharedCallCounterMap.getInstance();
     resolveState = new ResolveState(ldsResourceName);
+    ServerInfo serverInfo = xdsClient.getServerInfo(ldsResourceName);
```
Comment (ejona86) [inline xds/src/main/java/io/grpc/xds/XdsNameResolver.java:224]: This is not compatible with fallback. With fallback, each authority will have multiple server infos. We have to use the Args passed to the resource parser.

File: api/src/main/java/io/grpc/LoadBalancer.java [hunk 4]
```diff
@@ -552,6 +553,7 @@ public static final class PickResult {
     private final Status status;
     // True if the result is created by withDrop()
     private final boolean drop;
+    private String authorityOverrideHostname;
```
[the remark this answers is not in the document]
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/LoadBalancer.java]: Done.

File: api/src/main/java/io/grpc/LoadBalancer.java [hunk 5]
```diff
@@ -639,6 +651,18 @@ public static PickResult withSubchannel(
           false);
     }
 
+    /**
+     * Same as {@code withSubchannel(subchannel, streamTracerFactory)} but with an authority name
+     * to override in the host header.
+     */
+    public static PickResult withSubchannel(
```
Comment (ejona86) [inline api/src/main/java/io/grpc/LoadBalancer.java:658]: `@ExperimentalApi` here and below.

File: api/src/main/java/io/grpc/LoadBalancer.java [hunk 6]
```diff
@@ -682,6 +706,12 @@ public static PickResult withNoResult() {
       return NO_RESULT;
     }
 
+    /** Returns the authority override hostname if any. */
+    @Nullable
+    public String getAuthorityOverrideHostname() {
```
Comment (ejona86) [inline api/src/main/java/io/grpc/LoadBalancer.java:711]: Why is "hostname" is the method name? It basically never makes sense to say both authority and hostname.

File: api/src/main/java/io/grpc/LoadBalancer.java (hunk 5 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/LoadBalancer.java:659]: Done.

File: api/src/main/java/io/grpc/LoadBalancer.java (hunk 6 above)
Comment (kannanjgithub) [reply api/src/main/java/io/grpc/LoadBalancer.java]: Done.

**Review by ejona86 — APPROVED** [review]

**Pull request merged**